feat(action-bar): Add menu to ActionBar

RayRedGoose opened the pull request:
<!-- Thank you for your pull request, please provide a brief summary of what this introduces (mandatory). Please point out any code that may be non-obvious to reviewers by using in-code comments. -->

## Summary

Fixes: #1391  <!-- For bug fixes, use "Fixes". For new features use "Resolves". This helps link a PR to an issue and will show up in release notes. -->

This PR contains ActionBar refactoring to adding model functionality. It gives opportunity to add the specific for ActionBar overflow functionality: it shows only 3 first items even all action can be placed in ActionBar container and to add responsible functionality to hiding action that don't fit into ActionBar container based on container width. 

<img width="731" alt="Screen Shot 2022-05-16 at 11 28 09 AM" src="https://user-images.githubusercontent.com/48605821/168649500-f66acfe0-1bfa-4717-85fb-7e6a662a1feb.png">

<img width="538" alt="Screen Shot 2022-05-16 at 11 28 29 AM" src="https://user-images.githubusercontent.com/48605821/168649573-e3207678-987a-489b-a3b4-6a4c07e6c321.png">


<!-- This is the category in the release notes. Common categories are Components, Infrastructure, and Documentation -->
![category](https://img.shields.io/badge/release_category-Components-blue)

### Release Note
Optional release note message. Changelog and release summaries will contain a pull request title. This section will add additional notes under that title. This section is not a summary, but something extra to point out in release notes. An example might be calling out breaking changes in a labs component or minor visual changes that need visual regression updates. Remove this section if no additional release notes are required.

### BREAKING CHANGES
This PR contains breaking changes because using new ActionBar requires restructure old version and adding ActionBar.List as sub-component of ActionBar. This changes are covered by codemod but using an overflow behavior requires manual implementation.

---

## Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] tests are changed or added
- [ ] code has been documented
- [ ] design approved final implementation
- [ ] a11y approved final implementation
- [ ] code adheres to the [API & Pattern guidelines](https://workday.github.io/canvas-kit/?path=/story/welcome-dev-docs-api-pattern-guidelines--page)

## Additional References

<!-- Upload screenshots of the final component or any other artifacts that would help a reviewer understand the choices you made in the PR. -->

mannycarrera4 commented on modules/react/action-bar/stories/examples/Basic.tsx:
```diff
@@ -1,13 +1,16 @@
-import {PrimaryButton, SecondaryButton} from '@workday/canvas-kit-react/button';
-import {relatedActionsIcon} from '@workday/canvas-system-icons-web';
+import {TertiaryButton} from '@workday/canvas-kit-react/button';
+// import {relatedActionsIcon} from '@workday/canvas-system-icons-web';
```
```suggestion
```

mannycarrera4 commented on modules/react/action-bar/lib/useActionBarModel.tsx:
```diff
@@ -0,0 +1,74 @@
+
+  const menu = useMenuModel(
+    useMenuModel.mergeConfig(config.menuConfig as Required<typeof config.menuConfig>, {
+      id: `menu-${model.state.id}`,
```
do we want to prefix this with `action-bar-menu...` in case there's another menu in the dom?

RayRedGoose replied:
Nice catch! Thanks

mannycarrera4 commented on modules/react/action-bar/lib/ActionBarItem.tsx:
```diff
@@ -0,0 +1,56 @@
+   * index at the end. Only set this for advanced cases.
+   */
+  id?: string;
+  isPrimary?: boolean;
```
It's not possible to have a tertiary button or is it? Maybe I'd call this `isPrimaryButton`.

RayRedGoose replied:
Based on design spec and documentation, first button should be primary and others are secondary. So there is no tertiary buttons at all in Action Bar structure, but it's still possible to add  <TertiaryButton /> component as a child manually.

Agree with `isPrimaryButton`  👍

mannycarrera4 commented on modules/react/action-bar/lib/ActionBar.tsx:
```diff
@@ -1,37 +1,39 @@
+    OverflowButton: ActionBarOverflowButton,
+    /**
+     * The overflow menu of the ActionBar component. If there isn't enough room to render all the action
+     * items, the extra actions that don't fit will be overflowed into this menu.
```
I would add this to the `mdx` file as well. This tells the consumer how to use this. I think we should also mention what is the threshold of number of items before the the list is collapsed.

mannycarrera4 replied:
In order for this to work they have to add the prop `overflowButton` and also use the menu is that correct?

mannycarrera4 commented on modules/react/action-bar/lib/ActionBarList.tsx:
```diff
@@ -0,0 +1,71 @@
+   * </ActionBar.List>
+   */
+  children: ((item: T) => React.ReactNode) | React.ReactNode;
+  overflowButton?: boolean;
```
curious what others think, should we rename this something like `shouldOverflow` or `showOverflowButton`

mannycarrera4 commented:
@RayRedGoose this is looking really good! Just had a couple questions mostly. One thing I noticed is the overflow button would focus as part of the list of items when using the tab key. Should we use `useRovingIndex`? Looking more into it

RayRedGoose commented:
@mannycarrera4 Thank you for your notes. I've added documentation for Action Bar change. Now it covers model and steps for implementing responsive layout. Also, I fixed issue with order of the overflow button.

alanbsmith commented on modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx:
```diff
@@ -105,16 +105,53 @@ deprecated component now, you can reference
+    text: 'First Action',
+    onClick: () => console.log('first clecked')
+  }]
+});
 
```
Add a closing code block line here.
```suggestion
});
```
```

alanbsmith replied:
Add a closing code line block here

alanbsmith commented on modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx:
`````diff
@@ -105,16 +105,53 @@ deprecated component now, you can reference
+implement a responsive layout based on a container width
+([#1325](https://github.com/Workday/canvas-kit/pull/1585)).
+
+````
`````
```suggestion
```
```

mannycarrera4 commented on modules/react/action-bar/lib/ActionBar.tsx:
```diff
@@ -1,37 +1,39 @@
+  modelHook: useActionBarModel,
+  subComponents: {
+    List: ActionBarList,
+    Item: ActionBarItem,
```
I could go either way, but `item` is always a `button` maybe this was following tabs convention but should this we renamed to `ActionBar.Button` ?

```tsx
  <ActionBar>
     <ActionBar.List>
         <ActionBar.Button isPrimaryButton />
         <ActionBar.Button />
     </ActionBar.List>
   </ActionBar>
```

I'm just thinking out loud. Maybe `Item` makes sense because you're wrapping it in a `List`

mannycarrera4 replied:
this isn't a blocker BTW

RayRedGoose replied:
Yeah, `ActionBar.Item` have more sense to be used here, because it's a part of the list and also refers to action, not jut button. Also `Item` is closely connected with menu, so it's shows that `ActionBar.Item` is the same with items in the overflow menu.

RayRedGoose replied:
An also it refers to model state items

mannycarrera4 replied:
ok cool! make sense to me!

mannycarrera4 commented on modules/react/action-bar/lib/ActionBarItem.tsx:
```diff
@@ -0,0 +1,56 @@
+   * index at the end. Only set this for advanced cases.
+   */
+  id?: string;
+  isPrimaryButton?: boolean;
```
Is there a way for us to check what the first `item` is and automatically make it a `PrimaryButton` ?That way we don't even need this prop and the first item always passed is set to Primary, everything else is secondary? Maybe we can check model.state.items ? Again, thinking out loud, this is fine as it works!

RayRedGoose replied:
Model automatically makes the first item as a PrimaryButton. I'm not sure that it's reasonable to check items passed as children and change them to match a pattern "first item is a primary button". So it should be done manually as it's done for overflow functionality.

mannycarrera4 commented on modules/react/action-bar/lib/ActionBarItem.tsx:
````diff
@@ -0,0 +1,56 @@
+   * The contents of the action item. This will be the accessible name of the action for screen readers.
+   *
+   * ```tsx
+   * <ActionBar.Item>First Tab</ActionBar.Item>
````
```suggestion
   * <ActionBar.Item>First Button</ActionBar.Item>
```

RayRedGoose replied:
Thanks for catch!

mannycarrera4 commented on modules/react/action-bar/lib/useActionBarOverflow.tsx:
```diff
@@ -0,0 +1,46 @@
+      const isHidden = state.hiddenIds.includes(name);
+
+      const props = {
+        tabIndex: isHidden ? -1 : 0,
```
Button index? or tab as in tabbing?

mannycarrera4 commented on modules/react/action-bar/stories/ActionBar.stories.mdx:
````diff
@@ -21,21 +26,133 @@ yarn add @workday/canvas-kit-react
+it with its subcomponents using React context. It does not represent a real element.
+
+```tsx
+<ActionBar items={[]}>{/* Child components */}</Tabs>
````
```suggestion
<ActionBar items={[]}>{/* Child components */}</ActionBar>
```

mannycarrera4 commented on modules/react/action-bar/stories/ActionBar.stories.mdx:
````diff
@@ -21,21 +26,133 @@ yarn add @workday/canvas-kit-react
+
+```tsx
+<ActionBar.Item isPrimaryButton onClick={() => console.log('it's first')}>
+    First Tab
````
```suggestion
    First Button
```

mannycarrera4 commented on modules/react/action-bar/stories/examples/OverflowActionBar.tsx:
```diff
@@ -0,0 +1,51 @@
+        </ActionBar.Menu.Popper>
+      </ActionBar>
+      <hr />
+      <h4>Change tab container size</h4>
```
```suggestion
      <h4>Change Action Bar container size</h4>
```

mannycarrera4 commented on modules/react/action-bar/stories/ActionBar.stories.mdx:
```diff
@@ -4,12 +4,17 @@ import {ActionBar} from '@workday/canvas-kit-react/action-bar';
 
-Action Bar contains primary and secondary actions related to a page or task.
+Action Bar is a [compound component](/getting-started/for-developers/resources/compound-components/)
+contains primary and secondary actions related to a page or task.
```
```suggestion
that contains primary and secondary actions related to a page or task.
```

mannycarrera4 commented on modules/react/action-bar/stories/ActionBar.stories.mdx:
```diff
@@ -21,21 +26,133 @@ yarn add @workday/canvas-kit-react
+`ActionBar` includes a container `ActionBar` component and the following subcomponent:
+`ActionBar.List` which should contains `ActionBar.Item`.
+
+In a basic example of an `ActionBar` there is two buttons. The primary action button should be used
```
```suggestion
In a basic example of an `ActionBar` there are two buttons. The primary action button should be used
```

mannycarrera4 commented on modules/react/action-bar/stories/ActionBar.stories.mdx:
```diff
@@ -21,21 +26,133 @@ yarn add @workday/canvas-kit-react
+`ActionBar.List` which should contains `ActionBar.Item`.
+
+In a basic example of an `ActionBar` there is two buttons. The primary action button should be used
+only once and left aligned followed by secondary buttons. Tertiary Buttons should not be on the
```
```suggestion
only once and left aligned if content is left to right, followed by secondary buttons. Tertiary buttons should not be on the
```

mannycarrera4 commented on modules/react/action-bar/stories/ActionBar.stories.mdx:
```diff
@@ -21,21 +26,133 @@ yarn add @workday/canvas-kit-react
-## Responsive Behavior
+### Icons Example
+
+`ActionBar.Item` is a button, so it's possible to use other Button props with `ActionBar.Item` such
```
```suggestion
`ActionBar.Item` renders a `PrimaryButton` or `SecondaryButton`, so it's possible to use other Button props with `ActionBar.Item` such
```

mannycarrera4 commented on modules/react/action-bar/stories/ActionBar.stories.mdx:
```diff
@@ -21,21 +26,133 @@ yarn add @workday/canvas-kit-react
+### Overflow Example
+
+`ActionBar` container can contain up to 3 actions and an Overflow Menu if there are more than 3
+actions. 4th item and other remaining actions should be placed into an Overflow Menu that is
```
```suggestion
actions, the other remaining actions should be placed into an Overflow Menu that is
```

mannycarrera4 commented on modules/react/action-bar/stories/ActionBar.stories.mdx:
```diff
@@ -21,21 +26,133 @@ yarn add @workday/canvas-kit-react
+actions. 4th item and other remaining actions should be placed into an Overflow Menu that is
+launched by clicking the Overflow Button.
+
+To using an overflow behavior in ActionBar `showOverflowButton` prop should be passed to
```
```suggestion
To use the overflow behavior in ActionBar `showOverflowButton` prop should be passed to
```

mannycarrera4 commented on modules/react/action-bar/stories/ActionBar.stories.mdx:
```diff
@@ -21,21 +26,133 @@ yarn add @workday/canvas-kit-react
+launched by clicking the Overflow Button.
+
+To using an overflow behavior in ActionBar `showOverflowButton` prop should be passed to
+`ActionBar.List` and `ActionBar.Menu.Popper` structure should be placed inside `ActionBar`
```
```suggestion
`ActionBar.List` and `ActionBar.Menu.Popper` component should be used inside `ActionBar`
```

mannycarrera4 reviewed (approved):
wahooooooo
